amago/envs/builtin/xland_minigrid.py

```diff
@@ -0,0 +1,297 @@
+import random
+from collections import defaultdict
+from typing import Optional
+
+import gymnasium as gym
+import numpy as np
+
+try:
+    import jax
+    import xminigrid
+except ImportError:
+    jax = None
+    xminigrid = None
+else:
+    import jax.numpy as jnp
+    from jax import device_put as dp
+    from xminigrid.wrappers import (
+        DirectionObservationWrapper,
+        DmEnvAutoResetWrapper,
+        RulesAndGoalsObservationWrapper,
+    )
+
+from amago.envs import AMAGO_ENV_LOG_PREFIX
+
+
+def _swap_rules(old_goal, old_rule, old_tile, new_goal, new_rule, new_tile, replace):
+    goal = jnp.where(jnp.expand_dims(replace, 1), new_goal, old_goal)
+    rule = jnp.where(jnp.expand_dims(replace, (1, 2)), new_rule, old_rule)
+    tile = jnp.where(jnp.expand_dims(replace, (1, 2)), new_tile, old_tile)
+    return goal, rule, tile
+
+
+def _swap_trees(old_tree, new_tree, replace):
+    def select_fn(old, new):
+        diff = old.ndim - replace.ndim
+        axes = tuple([1 + i for i in range(diff)])
+        idxs = jnp.expand_dims(replace, axes)
+        return jnp.where(idxs, old, new)
+
+    return jax.tree_util.tree_map(select_fn, new_tree, old_tree)
+
+
+class XLandMiniGridEnv(gym.Env):
+    def __init__(
+        self,
+        parallel_envs: int,
+        k_shots: int = 25,
+        rooms: int = 4,
+        grid_size: int = 13,
+        ruleset_benchmark: str = "small-1m",
+        train_test_split: str = "train",
+        train_test_split_key: int = 0,
+        train_test_split_pct: float = 0.8,
+        jax_device: Optional[int] = None,
+    ):
+        assert (
+            jax is not None and xminigrid is not None
+        ), "JAX and xminigrid must be installed to use this environment"
+        self.rooms = rooms
+        self.grid_size = grid_size
+        self.k_shots = k_shots
+        self.ruleset_benchmark = ruleset_benchmark
+        self.parallel_envs = parallel_envs
+        self.jax_device = (
+            jax.devices()[jax_device]
+            if jax_device is not None
+            else jax.devices("cpu")[0]
+        )
+        print(f"Using JAX device {self.jax_device} for XLandMiniGridEnv")
+
+        # this always uses cuda memory for some reason?
```

Comment:
Yeah, this is just the default jax behaviour, i.e. putting all jax arrays to accelerator if available. You can put it to cpu manually. For example with `with jax.default_device(cpu_device):`.